[zexueteh] iP

zexueteh opened the pull request.

zexueteh committed:
feature: Introduce TARS Chatbot skeleton with hello & goodbye messages.

zexueteh committed:
feature: Implement Level-1 Echo functionality for TARS Chatbot

zexueteh committed:
fix: Echo functionality exits immediately after bye, chatbot reply formatting

zexueteh committed:
feature: Implement Level-2 store functionality for user inputs, display with list command

zexueteh committed:
feature: Implemented Task Class, modified list functionality, mark/unmark command to update task status

zexueteh committed:
refactor: Improve coding standards compliance for TARS chatbot

- Added Javadoc comments for all methods to enhance documentation.
- Replaced magic numbers with named constants. Example: maximum task size.
- Renamed variables and methods to follow Java naming conventions.
- Modularized code by extracting program logic into functions.
- Example: mark/unmark logic into 'MarkHandler'
- Improved code formatting and layout for better adherence to style guide.

zexueteh committed:
feature: Implemented Level-4 Todo, Event and Deadline functionality
- Task class added new Superclass method getSymbol and modified toString
- Added new classes Todo, Event and Deadline inherit from Task
- new method 'taskParser' that handles user input and creates Task Objects
- updated 'addTask' method to handle different class Types.
- Improved `printTaskList` formatting for better readability.
- Refactored task marking logic into `markHandler` to improve maintainability.
- Ensured basic input validation for command parsing.

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+import java.util.Scanner;
+
+public class TARS {
+    private static final String LINE_SEPERATOR = "    ____________________________________________________________";
```
Good job in using a constant to avoid a magic string

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+    private static final int MAX_TASK = 100;
+    private static Task[] taskList = new Task[MAX_TASK];
+
+    /**
```
Good job in using comments.

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+        String commandBody = line.substring(line.indexOf(" ") + 1);
+
+
+        if (!commandType.isEmpty() && (commandType.equals("todo") || commandType.equals("deadline") || commandType.equals("event"))) {
```
could split the conditional for better readability.

jing-yen replied:
Line length should be no longer than 120 chars.

jing-yen commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+import java.util.Scanner;
+
+public class TARS {
+    private static final String LINE_SEPERATOR = "    ____________________________________________________________";
+    private static int nTasks = 0;
+    private static final int MAX_TASK = 100;
+    private static Task[] taskList = new Task[MAX_TASK];
```
Great work using camelCase for variable names and SCREAMING_SNAKE_CASE for constants

jing-yen commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+import java.util.Scanner;
+
+public class TARS {
```
Class name should be in PascalCase

WayneCh0y replied:
Class name should be in Pascal case

icknee commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+     * - Marking Tasks as done: command "mark {task_number}"
+     * - Marking Tasks as undone: command "unmark {task_number}"
+     * - Exit: command "bye"
+     *
```
I like that you added user guidelines

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+            case "event":
+                newTask = new Event(description, to, from);
+                break;
+            }
```
Could consider using default block

icknee commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+                int index = parseMarkCommand(line);
+                boolean isMarking = line.startsWith("mark");
+                markHandler(index, isMarking);
+
```
some random empty lines

icknee commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+    private static void printHelloMessage() {
+        System.out.println(LINE_SEPERATOR);
+        System.out.println("      _____  _    ____  ____  ");
+        System.out.println("     |_   _|/ \\  |  _ \\/ ___| ");
```
the two empty lines on the top and bottom can be substituted with /n instead of those spaces

icknee commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+
+        System.out.println(LINE_SEPERATOR);
+        System.out.println("    Yes Captain. I've added this task.");
+        System.out.println("      " + taskList[nTasks]);
```
these indents can be subbed with /t

WayneCh0y commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String toString() {
+        return "[" + getSymbol() + "][" + (this.isDone ? "X" : " ") + "] " + this.description;
```
Could perhaps consider abstracting the (this.isDone ? "X" : " ") to a method

icknee reviewed (commented):
Clear and readable code! Keep it up!

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+
+        System.out.println(LINE_SEPERATOR);
+        System.out.println(
+            isMarking
```
Could consider abstracting this conditional into a function for better readability

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+                break;
+            } else if (line.equals("list")) {
+                printTaskList();
+            } else if (line.startsWith("mark") || line.startsWith("unmark")) {
```
perhaps could consider splitting this conditional for better readability.

WayneCh0y commented on src/main/java/TARS.java:
```diff
@@ -0,0 +1,190 @@
+
+        while (true) {
+            line = sc.nextLine().strip();
+            if (line.equals("bye")) {
```
maybe a switch case would be more suitable in this context

WayneCh0y reviewed (commented):
A clear and well documented code, keep it up!

zexueteh committed:
test: Implemented A-TextUiTesting for automated testing
- input.txt updated with commands corresponding to week 4 examples
- EXPTECTED.txt updated with expected output
- minor changes to TARS to remove trailing whitespaces

zexueteh committed:
refactor: Improve A-CodeQuality for TARS Chatbot.
- moved welcome/goodbye messages from multi-line prints to String constants
- replaced all instances of leading whitespace in prints to \t characters
- updated EXPECTED.txt to accommodate output changes
- reduced nesting in taskParser by splitting out switch statement
- Utilised abstract declaration for Task to prevent instantiation.

zexueteh committed:
refactor: Started on Level-5 error validation
- updated parseCommandType to be case-insensitive, throw exception for empty commands
- introduced CommandType Enum to validate string, eliminating magic strings.
- minor refactoring to other methods to restructure validation logic

zexueteh committed:
refactor: implemented custom A-Exceptions
- introduced new TARSInvalidCommandBodyException to validate command body
- created new parseCommandBody function to validate command body based on command type. LIST, MARK, UNMARK validation implemented
- updated parseCommandType to throw Exceptions
- Try-Catch block in main program loop that handles exceptions from two aforementioned parsing methods
- todo: finish validation for all command types. Implement command handler method to be called after validation.

zexueteh committed:
refactor: Implemented Level-5 error validation
- Implemented parseCommandBody to validate command arguments and enforce proper structure.
- Added validation for /by, /from, /to flags in DEADLINE and EVENT commands.
- Ensured MARK and UNMARK commands validate task index before processing.
- Refactored commandHandler() to improve separation of responsibility.

zexueteh committed:
refactor: Refined Level-5 Error Validation
- Updated TARSInvalidCommandBodyException constructor method to accept error message
- Implemented Stretch Goal: error-specific error messages for each Exception
- Small changes to validation logic, reordering of validation.
- input.txt updated with invalid command test cases
- EXPTECTED.txt updated with expected output for invalid cases

zexueteh committed:
Merge Level-5 features

zexueteh committed:
refactor: Restructured TARS into A-Packages
- Moved CommandType and TARSInvalidCommandBodException into TARS.command package
- Organized Task, Todo, Deadline and Event Classes into TARS.task package
- Moved TARS main class into TARS.ui package
- Updated all references to reflect new package structure
- Modified EXPECTED.txt and runtest.bat to align with updated structure

zexueteh committed:
Merge A-Packages restructuring

zexueteh committed:
refactor: Implemented taskList with A-Collections
- taskList implemented with ArrayList instead of Array
- updated array indexing with ArrayList Methods e.g. get, add
- nTasks can be dynamically represented with taskList.size()
- removed nTasks counter variable, updated logic in listing and validation

zexueteh committed:
feature: Implemented Level-6 delete function
- implemented new CommandType DELETE to Enum
- delete command implemented with same syntax/validation as mark/unmark
- created deleteHandler function that handles deletion logic
- Added DELETE case to commandHandler
- bug-fix to validation: amended search term for hasBy, hasFrom, hasTo flag detection booleans

zexueteh committed:
feature: Implemented Level-7 file storage function
- Added writeFile() method to save task list data to /data/data.txt file
- Implemented loadFile() method to read task list data from storage.
- checkDataPathExists() method ensures data directory/file exits before IO operations
- initializeTaskList() method converts stored data into Task objects saved in taskList array
- Refactored getTask() method to use by/to/from delimiters to parse String representation of Tasks stored in data file.

zexueteh committed:
merged Level-7 branch

zexueteh committed:
resolved conflict when merging Level-6 branch

sevenseasofbri commented on src/main/java/TARS/command/CommandType.java:
```diff
@@ -0,0 +1,14 @@
+        try {
+            return CommandType.valueOf(command.toUpperCase());
+        } catch (IllegalArgumentException e) {
+            // Modifying exception message
```
This comment does not sound like it is describing what is happening. You might have to be clearer.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+import java.io.FileWriter;
+
+public class TARS {
+    private static final String LINE_SEPERATOR = "\t____________________________________________________________";
+    private static final String HELLO_MESSAGE = """
+            \t  _____  _    ____  ____
+            \t |_   _|/ \\  |  _ \\/ ___|
+            \t   | | / _ \\ | |_) \\___ \\
+            \t   | |/ ___ \\|  _ < ___) |
+            \t   |_/_/   \\_\\_| \\_\\____/
+            \tHello, I'm TARS, your friendly chatbot assistant.
+            \tHow can I help you today?""";
+    public static final String GOODBYE_MESSAGE = "\tGoodnight Captain. Sleep well.";
```
Good use of constants.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+        printHelloMessage();
+        ArrayList<String> fileData = loadFile();
+        initializeTaskList(fileData);
+//        for (int i = 0; i < 4; i++) {
+//            System.out.println(fileData[i]);
+//        }
```
Avoid commenting out dead code. Remove it instead .

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+     * @return first word in line as commandType
+     */
+    private static CommandType parseCommandType(String line) throws IndexOutOfBoundsException, IllegalArgumentException {
+        // Check for empty line
```
This comment is not right before the relevant line, furthermore, you do not need to write comments for things that are obvious through the code.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+    }
+
+
+    /**
+
```
Ideally there should be no spacing like this in JavaDoc comments

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+        }
+        commandBody = line.substring(commandStartIndex);
+
+        boolean hasBy = commandBody.toLowerCase().contains("/by");
+        boolean hasFrom = commandBody.toLowerCase().contains("/from");
+        boolean hasTo = commandBody.toLowerCase().contains("/to");
```
Consider storing these magic literals in constants or vars instead.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+            }
+            break;
+
+        case MARK:
+        case UNMARK:
+        case DELETE:
```
Fallthrough statements in `switch-case` conditionals need to be explicitly succeeded by a `//Fallthrough` comment. 

> From the java [coding standard website](https://se-education.org/guides/conventions/java/basic.html#layout): The explicit // Fallthrough comment should be included whenever there is a case statement without a break statement.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+     * @return validated commandBody String
+     * @throws TARSInvalidCommandBodyException if the commandBody does not meet validation criteria
+     */
+    private static String parseCommandBody(String line, CommandType commandType)  throws TARSInvalidCommandBodyException {
```
This function is way over >30 LOC. Consider refactoring the internal levels of code into separate methods.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+        File dataFile = new File(dataFilePath);
+
+        String nextLine;
+
+        try {
+            Scanner fileScanner = new Scanner(dataFile);
+
+            while (fileScanner.hasNextLine()) {
+                nextLine = fileScanner.nextLine().trim();
+                if (nextLine.endsWith(")")) {
+                    nextLine = nextLine.substring(0, nextLine.length() - 1);
+                }
+
+                rawTaskData.add(nextLine);
+            }
```
There are ~3 levels of nesting here, refactoring the code to a method will improve readability.

sevenseasofbri commented on src/main/java/TARS/ui/TARS.java:
```diff
@@ -0,0 +1,559 @@
+            }
+
+        } catch (FileNotFoundException e) {
+            // file does not exist at the start
```
Consider removing this comment since it is not directly relevant to this segment of code.

sevenseasofbri reviewed (commented):
Good job so far! Well written JavaDocs. Your methods could be shorter though, so do look into that.

zexueteh committed:
refactor: Started on A-MoreOOP increment
- Separated functionality into command, logic, task and util packages
- new UI class to handle user interactions (e.g. console print, user input)
- new Storage class to handle file IO operations (only Load method implemented)
- new TaskList class to handle program memory (only initialise method implemented)
- new Parser class to parse file/command lines with regex and return Task/Command (only parse File implemented)
- new TaskType Enum to represent/validate Todo, Deadline, Event classes
- modifications to Task class and subclasses with default constructor, new TaskType attribute
- Todo: abstract Command and subclasses, parsing/validation/exception handling for command lines.

zexueteh committed:
refactor: Completed A-MoreOOP increment
- new Validator class validates user input before parsing. Throws TARSValidateCommandType/Param exceptions.
- Validator class validates 1. CommandType is valid 2. parameters of specific CommandType are valid.
- Parser parses input after validation, assumes valid input. Instantiates Command classes to be executed
- new Command Abstract class with execute method, which is inherited by specific Command subclasses e.g. AddCommand.
- TaskList CRUD methods and toString overwrite that returns all tasks as a contiguous String.
- Storage class write method implemented to write data to file.
- Updated text-ui-test script and expected outputs to reflect changes in project structure, UI.

zexueteh committed:
refactor: Improved Structure of Parser, Validator
- new RegexConstants class to store regex strings used by Parser and Validator
- renamed constant regex strings to follow SE-EDU convention

zexueteh committed:
feature: Implemented Level-8 DateTime replacement
- DateTimeParser class that handles parsing of DateTime objects from parameter string
- Replaces date strings in "yyyy-MM-dd" format to "d MMM yyyy" format (e.g. 2025-03-10 to 10 Mar 2025)
- Replaces time strings in "HHmm" format to "h.mm a" format (e.g 0127 to 1.27 am)
- Updated ParseTask to implement dateTime replacement when getting new command string

zexueteh committed:
feature: Implemented Level-9 find function
- updated CommandType with FIND command
- new FindCommand class that implements find functionality
- new /find searchterm command that matches the search term to any part of the Task string
- Updaded Parser/Validator classes to account for /find functionality

zexueteh committed:
refactor: A-JavaDoc added comments
- Added Javadoc comments to multiple classes and methods
- Documenting all non-private classes and methods, ensuring clarity and maintainability.

zexueteh committed:
Merge pull request #1 from zexueteh/branch-Level-8

Branch level 8

zexueteh committed:
Merge pull request #2 from zexueteh/branch-Level-9

feature: Implemented Level-9 find function

zexueteh committed:
merged Level-9 and JavaDoc branches

zexueteh committed:
Merge pull request #3 from zexueteh/branch-A-JavaDoc

Branch A-JavaDoc Merge

zexueteh committed:
Update TARS README.md

zexueteh committed:
Update README.md

zexueteh committed:
minor bug fix to parser

zexueteh committed:
Merge branch 'master' of https://github.com/zexueteh/ip

zexueteh committed:
Update README.md

formatting fix

zexueteh committed:
Add help command and minor bug fixes before release

zexueteh committed:
Merge branch 'master' of https://github.com/zexueteh/ip

zexueteh committed:
bug-fix: seperated parsing for index and descriptions

zexueteh committed:
bug-fix: Find command searches Task attributes only

zexueteh committed:
bug-fix: Datetime parser format updated to HH:mm

zexueteh committed:
Update README.md
